### PR TITLE
Support raw javascript in autoload and use autoload in the notebook

### LINKED ISSUE #3500: bokeh.js must load first, before widgets, compiler
refererence: https://groups.google.com/a/continuum.io/forum/#!topic/bokeh/aJPeJUottKk

ping @mattpap 

Any immediate thoughts? There are definitely ways to chain callbacks together to load one first. We'd have to special case the first to be the "load first" script, or else chain them all together in order in general. Or we could put a fixed number of timeout retries when `Bokeh` is not found as a stopgap.

### PR DESCRIPTION
**Pull request by mattpap**
This makes autoload support raw javascrip, te.g. logging through resources, not some custom code. Added an example showing that `autoload_static()` works (`embed/custom_server`). Notebook now uses autoload for initialization bokeh and rendering plots. Fixed `plugin-prelude.js` and `sever-dev`. There is some unfortunate code duplication around `AUTOLOAD_JS` template, but that can be improved in a follow-up PR. ~~Tests may fail at this point~~.

fixes #3500

### CODE REVIEW

**Comment by bryevdv** [comment]
I think my first inclination is to add the nice custom server example from this PR to #3520

**Comment by mattpap** [comment]
Forgot to mention that this makes custom models work in all cases for the first time.

**Comment by mattpap** [comment]
The benefit of this PR over #3520 is that it doesn't change how bokeh behaves by default, i.e. static HTML and server. Only autoload (static/server) and notebook are affected.

**Comment by bryevdv** [comment]
Can you check `bokeh/examples/embed/animated.py` with this PR?

**Comment by mattpap** [comment]
Fails with:
```WebSocket connection to 'ws://localhost:5006/ws?bokeh-protocol-version=1.0&bokeh-session-id=BaATC0nAexMIT7QljJvF7TWGRFJgsQecRhEbcA3T1THj' failed: Error during WebSocket handshake: Unexpected response code: 403
```
Seems unrelated, but I don't know.

**Comment by mattpap** [comment]
I'm getting that error on `master` as well.

**Comment by bryevdv** [comment]
@mattpap you need to allow CORS:
```
ERROR:bokeh.server.views.ws:Refusing websocket connection from Origin 'http://localhost:8000'; use --allow-websocket-origin=localhost:8000 to permit this; currently we allow origins False
```

File: bokeh/core/_templates/autoload_tag.html [hunk 1]
```diff
@@ -24,8 +24,6 @@
 <script
     src="{{ src_path }}"
     id="{{ elementid }}"
-    async="false"
```
Comment (havocp) [inline bokeh/core/_templates/autoload_tag.html]: heh, good catch.

File: bokeh/embed.py [hunk 2]
```diff
@@ -327,43 +329,32 @@ def autoload_static(model, resources, script_path):
...
-            modelid = model_id,
-            docid = doc_id,
-            loglevel = resources.log_level
-        )
-
-        return encode_utf8(js), encode_utf8(tag)
-
-def autoload_server(model, app_path="/", session_id=None, url="default", loglevel="info"):
+
+    script = _script_for_render_items(docs_json, render_items, wrap_script=False)
```
Comment (havocp) [inline bokeh/embed.py]: I like that we're now using `doc_js.js` for the autoload case

**Comment by bryevdv** [comment]
notebooks are not rendering in PhantomJS for this PR:

https://s3.amazonaws.com/bokeh-travis/0.11.0dev10-d4bc786/plotting/notebook/linked_panning.png

**Comment by mattpap** [comment]
I tested that example and it works as in `master`, i.e. renders properly, no errors in js console, but doesn't animate.

**Comment by bryevdv** [comment]
Hrm, it is animating for me.

**Comment by mattpap** [comment]
> notebooks are not rendering in PhantomJS for this PR:

I think I know the cause.

**Comment by bryevdv** [comment]
It animates on this PR branch for me too

**Comment by bryevdv** [comment]
the notebook above also renders correctly for me locally

**Comment by mattpap** [comment]
> you need to explicitly allow CORS:

btw. Examples should tell me that. Currently they say to simply run `bokeh serve`.

**Comment by mattpap** [comment]
> the notebook above also renders correctly for me locally

You mean running notebook locally by hand or locally in phantomjs (I presume the former)? phantomjs my fail because I used `global.Bokeh`, which evaluates to `this.Bokeh` and this scope may be different from `window`.

**Comment by bryevdv** [comment]
> btw. Examples should tell me that. Currently they say to simply run `bokeh serve`.

Yup,I had updated the README, did not realize it was in the docstring too. It's updated on the docs PR branch now.

**Comment by bryevdv** [comment]
> You mean running notebook locally by hand or locally in phantomjs (I presume the former)?

yes by hand, sorry the other possibility had not occurred to me.

**Comment by mattpap** [comment]
I'm not able to test notebooks at all with phantomjs locally (not sure why, no errors anywhere). Lets see what travis ci says on changing `{global->window}.Bokeh`.

**Comment by havocp** [comment]
I'm not sure what we want to do today on a timeline. Given a little more time I'd really like to combine aspects of both patches. What I like here:

 * using doc_js.js for the autoload case as well
 * delete `async="false"` junk in html
 * fix server-dev
 * fix name of css_files to css_urls

What I like in #3520:

 * break up autoload_js.js so the bootstrap function is in its own file, avoiding a multi-purpose template (that may or may not have elementid for example). autoload_js.js is only for autoload, bootstrap is the thing shared with notebook (well, with everything, as I did it, but we could remove it from js_resources if we also remove it from doc_js and then wrap doc_js later only for the autoload case as you did)
 * public API access for bokeh_load, vs. private to autoload_js.js, which means one can write a script tag that works in both autoload and not scenarios. that's used for example to show "loaded successfully" with all the logic for that right in notebook_load.html rather than having to put it in the js_raw; it might also be needed if we wanted to keep loglevel working in autoload case (?)
 * bokeh_load still exists/works for the inline case, it just doesn't have to actually load stuff
 * only shows "loaded successfully" after it actually has

Not sure:
 * should we drop loglevel param from autoload_server or keep it working?

Good in both:
 * get rid of _wrap_in_function
 * fixes #3500 

Could cleanup for both:
 * we are doing this `function(global)` thing but then using `window` instead of `global`, or maybe I don't understand this fully, but I would think we should use global instead of window or don't wrap in `function(global)`

The main conceptual difference between the patches I think is that you are defining a rule that all "needs window.Bokeh to exist" code has to be known in advance - it's either passed to AUTOLOAD_JS as raw JS so it's deferred appropriately, or it assumes we are not using dynamic load and that Bokeh has been included in the original HTML of the page. There's no way to "hook" the autoload at a later time (after AUTOLOAD_JS has rendered, or from JavaScript), because there's no exposed bokeh_load. But on the other hand of course, you've also removed all situations from our own codebase that need to use bokeh_load.

I don't know what we should do today.

**Comment by bryevdv** [comment]
I very much like the `load_bokeh` approach in the other patch, and would like to pursue that, *at some point*. I think for today we have to come to a decision in the next hour or so. It is imperative that we get out an RC1 for people to test with.

**Comment by mattpap** [comment]
Looking at the report it seems that none of server examples work, but they do work locally. Maybe `tests/travis/test` needs updating after some recent changes (?).

**Comment by bryevdv** [comment]
It's a test problem, I forgot to fix it, they are connecting with the wrong session id.

**Comment by havocp** [comment]
@bryevdv to me if you want something in the next hour you could pick either PR and it should fix #3500 . This one doesn't add new bokeh_load API but does remove autoload_server(loglevel) API. So one way to decide is to think about which of those API changes you're happier with, and then we can do some porting from one PR to the other post-release.

**Comment by mattpap** [comment]
> only shows "loaded successfully" after it actually has

This is something I wanted to add here, but didn't want to postpone submitting a PR.

**Comment by mattpap** [comment]
> but does remove autoload_server(loglevel) API

Now `loglevel` is maintained uniformly at `resources` level. Or do I miss a point?

**Comment by mattpap** [comment]
> we are doing this function(global) thing but then using window instead of global

This definitively needs improvement and more systematic approach, but is one of many things I didn't want to touch at this point.

**Comment by bryevdv** [comment]
Let's merge this one then, and update parts from the other PR after the release. @mattpap can you merge this as soon as you are satisfied with it?

**Comment by mattpap** [comment]
> I very much like the load_bokeh approach in the other patch

I didn't properly review the other PR, but my ad-hoc worry is about development implications and how cleanly this embeds in third-party sites. Originally I didn't want to even expose global `Bokeh` to a web page, but this idea didn't survive very long.

**Comment by havocp** [comment]
sounds good.

**Comment by mattpap** [comment]
@bryevdv, phantomjs still fails. I'm trying to debug this locally, but I'm getting various erroneous behavior from phatomjs + jupyter (e.g. error 426, @damianavila). Investigating what's wrong here. However, I tested all examples locally and they work, so I'm hoping this is just a testing issue.

**Comment by bryevdv** [comment]
@mattpap I just noticed that as well. I do think we should not let that hold up merging it if examples are working locally. Let's shoot to merge this by 1:30 CST to have an RC1 available by 2pm CST

**Comment by mattpap** [comment]
OK, I don't plan to add anything to this PR. I will be focusing on resolving phantomjs issue. If there are no new objections, then lets merge.

**Comment by damianavila** [comment]
> I'm trying to debug this locally, but I'm getting various erroneous behavior from phatomjs + jupyter (e.g. error 426, @damianavila)

Weird... looking into the web, that error seems related with some websocket incompatibility: http://tools.ietf.org/html/rfc6455#section-4.2.2

**Comment by bryevdv** [comment]
@mattpap I am going to merge the other tiny PRs (map bounds, box justify, etc) first, then this one.

**Comment by mattpap** [comment]
> However, I tested all examples locally and they work, so I'm hoping this is just a testing issue.

The actual reason was that notebook embed code didn't allow `INLINE` resources, which are used by our test infrastructure when testing notebooks. PR #3532 resolves this issue. On a side note, don't depend on notebook test results on travis ci, because most of the time it reports success in this case. This is because notebook does't propagate most of errors to the top-level.

**Pull request merged**